internal/impl/git/input.go

```diff
@@ -0,0 +1,636 @@
+	// lastCommit is the hash of the most recently processed commit.
+	lastCommit plumbing.Hash
+	// lastCommitMu is a lock for accessing lastCommit.
+	lastCommitMu sync.RWMutex
+	// tempDir is the temporary directory where the repository is cloned.
+	tempDir string
+	// mgr is the service resources manager.
+	mgr *service.Resources
+}
+
+// fileEvent represents a file change event.
+type fileEvent struct {
+	// path is the absolute path to the file.
+	path string
+	// isDeleted indicates whether the file was deleted.
+	isDeleted bool
+}
+
+// init registers the Git input plugin with the service registry.
+func init() {
+	err := service.RegisterInput(
+		"git", gitInputConfig(),
+		func(parsedCfg *service.ParsedConfig, mgr *service.Resources) (service.Input, error) {
+			conf, err := inputCfgFromParsed(parsedCfg)
+			if err != nil {
+				return nil, err
+			}
+
+			return newInput(conf, mgr)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// newInput creates a new Git input instance from a parsed configuration.
+func newInput(cfg inputCfg, mgr *service.Resources) (*input, error) {
+	i := &input{
+		cfg:       cfg,
+		filesChan: make(chan fileEvent),
+		shutSig:   shutdown.NewSignaller(),
+		log:       mgr.Logger(),
+		mgr:       mgr,
+	}
+
+	return i, nil
+}
+
+// Connect implements service.Input. It initializes the Git repository by creating
+// a temporary directory, cloning the repository, and starting the polling routine.
+func (in *input) Connect(ctx context.Context) error {
+	// Create a temporary directory for the repository
+	tmpDir, err := os.MkdirTemp("", "git-input-*")
+	if err != nil {
+		return fmt.Errorf("failed to create temp directory: %w", err)
+	}
+	in.tempDir = tmpDir
+
+	// If checkpoint cache is configured, try to get the last processed commit
+	var cachedCommitHash string
+	if in.cfg.checkpointCache != "" {
+		if err := in.mgr.AccessCache(ctx, in.cfg.checkpointCache, func(cache service.Cache) {
+			lastCommitBytes, cacheErr := cache.Get(ctx, in.cfg.checkpointKey)
+			if cacheErr != nil && !errors.Is(cacheErr, service.ErrKeyNotFound) {
+				err = fmt.Errorf("failed to get last commit from cache: %w", cacheErr)
+				return
+			}
+			cachedCommitHash = string(lastCommitBytes)
+		}); err != nil {
+			return err
+		}
+
+		if cachedCommitHash != "" {
+			in.log.Infof("continuing from cached last commit: %q", cachedCommitHash)
+		}
+	}
+
+	// Clone the repository
+	if err := in.cloneRepo(ctx); err != nil {
+		_ = os.RemoveAll(tmpDir)
+		return fmt.Errorf("failed to clone repo: %w", err)
+	}
+
+	// Start polling for changes
+	go in.pollChanges(cachedCommitHash)
+
+	return nil
+}
+
+// Read implements service.Input. It returns the next available file content as a message,
+// or returns an error if the context is cancelled or shutdown is signaled.
+func (in *input) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			in.shutSig.TriggerHasStopped()
```

Comment:
This is annoying, but Close() is not called while Read is still blocking.

Reply:
This feels wrong. `TriggerHasStopped` should be called when `pollChanges` goroutine finishes meaning that everything has stopped and shutdown before Close exits.